Source/DataSources/KmlExporter.js

```diff
@@ -0,0 +1,1165 @@
+    '../Core/TimeInterval',
+    '../Core/TimeIntervalCollection',
+    '../Scene/HeightReference',
+    '../Scene/HorizontalOrigin',
+    '../Scene/VerticalOrigin',
+    '../ThirdParty/when'
+], function(
+    BillboardGraphics,
+    CompositePositionProperty,
+    ModelGraphics,
+    SampledPositionProperty,
+    SampledProperty,
+    ScaledPositionProperty,
+    RectangleGraphics,
+    Cartesian2,
+    Cartesian3,
+    Cartographic,
+    Color,
+    createGuid,
+    defaultValue,
+    defined,
+    defineProperties,
+    Ellipsoid,
+    isArray,
+    Iso8601,
+    JulianDate,
+    CesiumMath,
+    Rectangle,
+    ReferenceFrame,
+    Resource,
+    RuntimeError,
+    TimeInterval,
+    TimeIntervalCollection,
+    HeightReference,
+    HorizontalOrigin,
+    VerticalOrigin,
+    when) {
+        'use strict';
+
+        var BILLBOARD_SIZE = 32;
+        var kmlNamespace = 'http://www.opengis.net/kml/2.2';
+        var gxNamespace = 'http://www.google.com/kml/ext/2.2';
+        var xmlnsNamespace = 'http://www.w3.org/2000/xmlns/';
+
+        function ExternalFileHandler(modelCallback) {
+            this._files = {};
+            this._promises = [];
+            this._count = 0;
+            this._modelCallback = modelCallback;
+        }
+
+        ExternalFileHandler.prototype.texture = function(texture) {
+            if (typeof texture === 'string') {
+                return texture;
+            }
+
+            if (texture instanceof Resource) {
+                return texture.url;
+            }
+
+            if (texture instanceof HTMLCanvasElement) {
+                var deferred = when.defer();
+                this._promises.push(deferred.promise);
+
+                var filename = 'texture_' + (++this._count) + '.png';
+                var that = this;
+                texture.toBlob(function(blob) {
+                    that._files[filename] = blob;
+                    deferred.resolve();
+                });
+
+                return filename;
+            }
+
+            return '';
+        };
+
+        ExternalFileHandler.prototype.model = function(model, time) {
+            var modelCallback = this._modelCallback;
+            if (!defined(modelCallback)) {
+                throw new RuntimeError('Model callback must be specified if there are models in the entity collection');
+            }
+
+            return modelCallback(model, time);
+        };
+
+        defineProperties(ExternalFileHandler.prototype, {
+            /**
+             * Resolves when all external files are ready
+             * @memberof GlExternalFileHandlerobe.prototype
+             * @type {Promise}
+             */
+            promise : {
+                get : function() {
+                    return when.all(this._promises);
+                }
+            },
+            /**
+             * An object with filename and blobs for external files
+             * @memberof ExternalFileHandler.prototype
+             * @type {ObjectId}
```

Comment:
Incorrect type.

Reply:
Actually, does this even need doc?  Same as the promise before it, isn't this an internal class?